### PR TITLE
DrivAerML: Push width further — 4L/640d

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
Width scaling has been consistently beneficial: 4L/256d→4L/320d→4L/512d each improved. 4L/640d pushes width further. With 96GB VRAM and batch_size=1, it should fit. The key question is whether the per-epoch time increase is offset by the capacity gain.

## Instructions
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 640 \
  --model-heads 10 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --agent zenitsu \
  --wandb-name "zenitsu/drivaerml-4L640d" \
  --wandb-group "drivaerml-width-scaling"
```

Note: 640/64=10 heads for consistent dim-per-head. If OOM, try reducing --drivaerml-train-surface-points to 40000.

Report val_primary/surface_rel_l2_pct, best epoch, total epochs, time/epoch, and peak VRAM usage.

## Baseline
Current best: **4.619%** at epoch 256 (4L/512d, 267 epochs, 180-min)
External target: <3.71%

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT zenitsu:

## Results

### DrivAerML 4L/640d Width Scaling

| Metric | Baseline (4L/512d) | This Run (4L/640d) | External Target |
|--------|-------------------|---------------------|-----------------|
| val_primary/surface_rel_l2_pct | **4.619%** (epoch 256) | **13.05%** (epoch 47) | <3.71% |
| Final val_primary (end of run) | — | 54.49% (diverged) | — |
| Test primary | — | 56.07% (diverged checkpoint) | — |

**Total epochs**: 266 (180-min timeout)
**Peak VRAM**: 7,577 MiB / 97,887 MiB (GPU 0)
**W&B run**: [`0s1jpa2s`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/0s1jpa2s)

**Command used** (exactly as specified in PR):
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 640 \
  --model-heads 10 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --agent zenitsu --wandb-name "zenitsu/drivaerml-4L640d" \
  --wandb-group "drivaerml-width-scaling"
```

### What happened

**Clear negative result.** The 4L/640d model is fundamentally unstable with this cosine LR restart schedule (lr=5e-4, cosine-t-max=30).

**Training instability timeline:**
- **Epochs 1–47**: Healthy convergence, best metric 13.05% at epoch 47. Loss and grad norms normal (~1.0).
- **Epoch 48**: First instability spike — val jumped from 13% to 52%, grad norm 3.3x. Model recovered over ~6 epochs.
- **Epoch 55**: Second spike — val to 63%. Recovered but slower.
- **Epoch 62**: Third spike — val to 60%. Recovery even slower, never fully returned to pre-spike best.
- **Epoch 88**: Fourth spike — val to 53%.
- **Epoch 91**: Fifth spike — grad norm exploded to 23.8 (vs normal ~1.0).
- **Epoch 99**: Terminal divergence — grad norm hit 510, then escalated to millions/billions. Model never recovered.
- **Epochs 99–266**: Completely diverged. Final grad norm: 8.4 billion.

**Key observation:** The instability pattern is tied to cosine LR restarts. The wider model (640d) is far more sensitive to the LR jumps at cycle boundaries than the 512d model was. Each spike caused lasting damage, and by epoch 99 the accumulated instability became unrecoverable.

**Best metric (13.05%) is 2.8x worse than the 4.619% baseline** — the wider model not only fails to improve on 512d, it can't even come close. This suggests width scaling has hit diminishing returns or a stability wall at 640d with the current training recipe.

### Suggested follow-ups

1. **Lower LR for 640d**: If 640d is still worth exploring, try lr=2e-4 or 1e-4 to reduce sensitivity to cosine restarts.
2. **Gradient clipping**: Adding grad norm clipping (e.g. max_norm=1.0) could prevent the cascading instability.
3. **Longer cosine-t-max**: cosine-t-max=60 or 90 would reduce the frequency and magnitude of LR jumps.
4. **Warmup restarts**: Smooth cosine warmup restarts instead of hard jumps at cycle boundaries.
5. **Conclusion**: Width scaling beyond 512d requires a fundamentally different LR/training strategy. The current recipe (lr=5e-4, cosine-t-max=30) is tuned for 512d and breaks at 640d.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2805 because ADVISOR: Closing — your 4L/640d result of 13.055% diverged badly, far from the 4.619% baseline. Wider architectures without stability controls cause divergence on DrivAerML. Noted for future experiments, zenitsu..
